### PR TITLE
Assignment4: jdriscoll1 and alowe4

### PR DESCRIPTION
**Pull request by jdriscoll1**
Extra Credit Attempted

### CODE REVIEW

**Review by Nolan-Posey — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Everything looks great! The only thing I didn't see was some tests for value equality for the Entities Book, Student, and Employee, but I did see your tests for the FullName record. 

My only suggestions would be:
1.) To add a simple value equality test for Book, Student, and Employee.
2.)  This is totally optional, but instead of having to make a test class for Book, Student, and Employee, I would recommend making a "RecordsTests" test class and then adding all your value equality tests to it. You could even move everything from your FullNameTests class over there if you so choose just to get it all into one file.

Overall, good job!
-Nolan

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -1,24 +1,26 @@
-﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+﻿
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 using System;
 using System.Collections.Generic;
 
+[assembly: CLSCompliant(true)]
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:6]: Why add this to your tests?

File: Logger.Tests/FileLoggerTests.cs [hunk 2]
```diff
@@ -6,19 +6,19 @@ namespace Logger.Tests;
 public class FileLoggerTests : FileLoggerTestsBase
 {    
     [TestMethod]
-    public void Create_GivenClassAndValidFileName_Success()
+    public void CreateGivenClassAndValidFileNameSuccess()
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:9]: You can keep the underscores in your test names, this is the typical way to name unit tests in c#

File: Logger.Tests/RecordTests.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            Employee e1 = new(name, "Programmer", "Walmart");
+            Employee e2 = new(name, "Programmer", "Walmart");
+            Assert.AreNotEqual<Employee>(e1, e2);
+
+
+        }
+
+
+        [TestMethod]
+        public void BooksAreNotEqual()
```
Comment (mmwoodfo) [inline Logger.Tests/RecordTests.cs:63]: You should make sure to check if equal as well!

File: Logger/Records.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    // that the two objects are tied to one object
+
+    /* MUTABLE OR IMMUTABLE EXPLANATION:  */
+    // Should the full name structure be mutable or immuable? Immutable
+    // There are three reasons that the Full Name record should be immutable
+    // Firstly, whenever we change the name, we should be creating a new Full Name Object rather than changing
+    // a previous one for better clarity
+    // Secondly, it should be immuatable so that it does not change a different object elsewhere in the code
+    // Thirdly, hash codes are calculated from the direct data storage, if we chagne the hash code we could lose information
+    public readonly record struct FullName(string FirstName, string LastName, string? MiddleName = null)
```
Comment (mmwoodfo) [inline Logger/Records.cs:25]: You should split these records into their own files.

File: Logger/IEntity.cs [hunk 5]
```diff
@@ -1,6 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger
 {
+    public interface IEntity
+    {
+        public Guid Id { get; init; }
```
Comment (mmwoodfo) [inline Logger/IEntity.cs:12]:
```suggestion
        Guid Id { get; init; }
```

File: Logger/IEntity.cs [hunk 6]
```diff
@@ -1,6 +1,17 @@
...
+        public string Name { get; }
```
Comment (mmwoodfo) [inline Logger/IEntity.cs:13]:
```suggestion
        string Name { get; }
```

File: Logger/BaseLogger.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@
 public abstract class BaseLogger
 {
     public string LogSource { get; }
-    public BaseLogger(string logSource) => LogSource = string.IsNullOrWhiteSpace(logSource)
+    protected BaseLogger(string logSource) => LogSource = string.IsNullOrWhiteSpace(logSource)
```
Comment (mmwoodfo) [inline Logger/BaseLogger.cs:8]: Why make this protected?

File: Logger/AbstractRecords.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger
+{
+ public abstract record class Entity : IEntity
```
Comment (mmwoodfo) [inline Logger/AbstractRecords.cs:9]: Your file name should reflect this class name.

**Pull request closed without merging**